net: lwm2m: dont use system workqueue for services

mike-scott opened the pull request:
"It's a Trap!" -- Admiral Ackbar

When moving to the BSD-socket APIs, the original thread running LwM2M
periodic services such as observes and lifetime updates, was replaced
with a re-occuring workqueue job.  To save the overhead of creating a
new thread, I used the system workqueue for these jobs.

This was a mistake.  If these jobs hit a semaphore or wait for some
reason, it cannot be prempted due to the priority of the system work
queue.

Let's instead add this service handling to the thread that we already
use for polling sockets.  This also removes a configuration issue where
the system workqueue stack size needed to be increased.  This can now
be adjusted via the LWM2M_ENGINE_STACK_SIZE knob.

Directly fixes semaphore usage in the socket-based DNS code.
This was introduced as a bugfix for non-responsive DNS server hanging
the Zephyr device forever.  However, this probably fixes randomly
seeming hangs on the device.


Fixes  #15204 

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: dont use system workqueue for services

"It's a Trap!" -- Admiral Ackbar

When moving to the BSD-socket APIs, the original thread running LwM2M
periodic services such as observes and lifetime updates, was replaced
with a re-occuring workqueue job.  To save the overhead of creating a
new thread, I used the system workqueue for these jobs.

This was a mistake.  If these jobs hit a semaphore or wait for some
reason, it cannot be prempted due to the priority of the system work
queue.

Let's instead add this service handling to the thread that we already
use for polling sockets.  This also removes a configuration issue where
the system workqueue stack size needed to be increased.  This can now
be adjusted via the LWM2M_ENGINE_STACK_SIZE knob.

Directly fixes semaphore usage in the socket-based DNS code.
This was introduced as a bugfix for non-responsive DNS server hanging
the Zephyr device forever.  However, this probably fixes randomly
seeming hangs on the device.

Signed-off-by: Michael Scott <mike@foundries.io>

andyross commented:
Work queue jobs really should never be blocking for any reason, that's an abuse of the API, the whole point of which is to isolate units of actual **work** so they can be scheduled asynchronously.   Even if you don't run into priority problems doing this, you are starving the other users of the queue.

Seems like an OK workaround, but this isn't a root cause fix.  Maybe we should have some kind of assertion layer to catch behavior like this, maybe flag a thread as  "must stay schedulable" while the work_q callback is in progresss...

jukkar reviewed (approved):
LGTM.

Do we actually need to have the #15210 as the DNS resolver already has timeout mechanism which should work just fine. So the fix in #15210 is something that we should not normally need in the socket layer.

pieterjanc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3810,16 +3806,12 @@ static void lwm2m_engine_service(struct k_work *work)
 		/* service is due */
 		if (timestamp > service_due_timestamp) {
```
If timestamp == service_due_timestamp, the work will not be handled. engine_next_service_timeout_ms will return 0 and lwm2m_engine_service will be called repeatedly without the work being handled.
```suggestion
		if (timestamp >= service_due_timestamp) {
```

mike-scott replied:
@pieterjanc I will add this as a second patch.  Thank you for pointing out the zero case.

mike-scott commented:
@jukka This is a companion PR to the Socket-based DNS fix.  Both are needed.

mike-scott commented:
@jukkar Should I add the fix for @pieterjanc issue here or do it as a separate PR?

jukkar commented:
> Should I add the fix for @pieterjanc issue here or do it as a separate PR?

Just add it to this PR.

mike-scott committed:
net: lwm2m: fix periodic services handling

This fixes an issue where if timestamp == service_due_timestamp,
we don't call the periodic service.  Then the following call to
engine_next_service_timeout_ms() returns 0 because the service
is still due and lwm2m_engine_service() is called again.
This process repeats several times until the value of
k_uptime_get() changes and then the work is finally handled.

Previously, the resolution of k_uptime_get() was in ms.  A recent
change to this API defaults Zephyr so that the resolution is
set via CONFIG_SYS_CLOCK_TICKS_PER_SEC (default 100).

This means the value of k_uptime_get() only changes every 10ms.

Reported-by: Github User pieterjanc
Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
@pieterjanc I've added a new patch to this PR to address the issue you brought up.  I explain the issue in detail in the commit description.   Please make sure my understanding of this issue is correct.

pieterjanc commented:
Perfect! This will also fix #15204 which I opened just before this pull request.
Thanks!
